recipes/gst-plugins-base/all/conanfile.py

```diff
@@ -0,0 +1,217 @@
+from conans import ConanFile, tools, Meson, VisualStudioBuildEnvironment
+from conans.errors import ConanInvalidConfiguration
+import glob
+import os
+import shutil
+
+
+class GStPluginsBaseConan(ConanFile):
+    name = "gst-plugins-base"
+    description = "GStreamer is a development framework for creating applications like media players, video editors, " \
+                  "streaming media broadcasters and so on"
+    topics = ("gstreamer", "multimedia", "video", "audio", "broadcasting", "framework", "media")
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://gstreamer.freedesktop.org/"
+    license = "GPL-2.0-only"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "with_libalsa": [True, False],
+        "with_gl": [True, False],
+        "with_introspection": [True, False],
+        }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+        "with_libalsa": True,
+        "with_gl": False,
+        "with_introspection": False,
+        }
+    _source_subfolder = "source_subfolder"
+    _build_subfolder = "build_subfolder"
+    exports_sources = ["patches/*.patch"]
+
+    generators = "pkg_config"
+
+    @property
+    def _is_msvc(self):
+        return self.settings.compiler == "Visual Studio"
+
+    def validate(self):
+        if self.options.shared != self.options["gstreamer"].shared or \
+            self.options.shared != self.options["glib"].shared:
+                # https://gitlab.freedesktop.org/gstreamer/gst-build/-/issues/133
+                raise ConanInvalidConfiguration("GLib, GStreamer and GstPlugins must be either all shared, or all static")
+        if tools.Version(self.version) >= "1.18.2" and\
+           self.settings.compiler == "gcc" and\
+           tools.Version(self.settings.compiler.version) < "5":
+            raise ConanInvalidConfiguration(
+                "gst-plugins-base %s does not support gcc older than 5" % self.version
+            )
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+        self.options['gstreamer'].shared = self.options.shared
```

Comment:
Should this be changed to a check in `validate`?
```python
if self.options.shared and not self.options["gstreamer"].shared:
    raise ConanInvalidConfiguration("The options gstreamer:shared and {}:shared must be the same".format(self.name))
```

Reply:
validate already has this check.

see above on line 42: https://github.com/conan-io/conan-center-index/pull/7142/files#diff-65119968a00cace4db47857f3233cdca7f2de2e0f3ff99cbd57a8fae7a271682R42

Reply:
Indeed. Thanks.
But this line overrides user arguments.
What happens if  I build with the following arguments?
```
conan create . gst-plugins-base/1.19.1@ -o gstreamer:shared=True -o gst-plugins-base:shared=False
conan create . gst-plugins-base/1.19.1@ -o gstreamer:shared=False -o gst-plugins-base:shared=True
```

Reply:
the bad things happen for sure.
GLib is designed to be present only once within an entire address space.
mixing shared/static leads to undefined behavior (e.g. CI hands).
literally, we run into the issue https://gitlab.freedesktop.org/gstreamer/gst-build/-/issues/133
e.g. locally I get these errors:
```
objc[21642]: Class GNotificationCenterDelegate is implemented in both /Users/sse4/.conan/data/gstreamer/1.19.1/_/_/package/456f111003e278e1a2f9a487e4964b0dec0b5ce3/lib/libgstnet-1.0.0.dylib (0x100a663b8) and /Users/sse4/.conan/data/gst-plugins-base/1.19.1/_/_/package/15f899debcee0aa9e4e9a72e52b66594fa545dbe/lib/libgstrtsp-1.0.0.dylib (0x1004c2378). One of the two will be used. Which one is undefined.
```
and many errors of that type:
```
(process:3125): GLib-GObject-CRITICAL **: 15:57:26.899: g_param_spec_pool_lookup: assertion 'pool != NULL' failed
```
these errors indicate two instances of glib exist in the address place.

the bad things also happen with conan side here...
I gonna submit new bug, but I doubt it will be ever fixed (at least in 1.x).
if I request the following build:
```
conan create . gst-plugins-base/1.19.1@ \
 -o gst-plugins-base:shared=True \
 -o gstreamer:shared=True \
 -o glib:shared=True \
```
what do you think happens?
I am requesting all `glib`, `gstreamer` and `gst-plugins-base` to be shared libraries.
does conan guarantee that all of them will be shared?
nope, it surprisingly consumes `gstreamer:shared=True` linked to `glib:shared=False`.
and there seems to be no way to override it from command line?

why does it happen?
both packages `gstreamer:shared=True` with `glib:shared=True` and `gstreamer:shared=True` with `glib:shared=False` have the same package id!
for me, running on Mac:
```
conan info gstreamer/1.19.1 -o gstreamer:shared=True -o glib:shared=False
conan info gstreamer/1.19.1-o gstreamer:shared=True -o glib:shared=True
```
I get the exactly same result:
```
gstreamer/1.19.1
    ID: 5a1e3781444d60ee53a8d7e762ece8f559cf15ca
```
so the options `glib:shared` is erased from the package id of the `gstreamer` package.

are these two packages supposed to have the same package id? I don't think so.
the first one embeds glib objects into the gstreamer DLL.
the second one links gstreamer DLL into glib DLL.
I think that's entirely different package with a different behavior.

Reply:
Isn't the problem you're describing the need for full transitive package ids? `default_package_id_mode` and `full_transitive_package_id`?
But you're right, all shared libraries built by c3i have all static libraries built-in. This is really awkward.

My comment was about the recipe overriding user options, which is a no no imho.

Perhaps conan needs a`post_validate` method.
A method in which a recipe can check the complete dependency tree.
Then , the following can be added to the `glib` recipe:
```python
def post_validate(self):
    if not self.options.shared:
        consumers = self.tree.get_consumers(self.name)
        shared_consumers = [consumer for consumer in consumers if getattr(consumer.options, "shared", False) == True]
        if len(shared_consumers) > 1:
            raise ConanInvalidConfiguration("glib is used by more then 1 library that is built as a shared library. This will lead to undefined behavior.")
```

Reply:
I don't think we need `post_validate`.
`validate` already runs for the final dependency tree, after all overrides computed.
```
Another important difference with the configure() method, is that validate() is evaluated after the graph has been computed and the information has been propagated downstream. So the values used in validate() are guaranteed to be final real values, while values at configure() time are not.
```

Reply:
Indeed. I thought about it too late.